server/src/main/java/io/crate/statistics/ReservoirSampler.java

```diff
@@ -89,20 +93,59 @@ public final class ReservoirSampler {
     private CircuitBreakerService circuitBreakerService;
     private final IndicesService indicesService;
 
+    private volatile ByteSizeValue readLimit;
+    private volatile ByteSizeValue updatedReadLimit;
+    private volatile boolean analyzing;
+
+    private final RateLimiter rateLimiter;
+    private long bytesSinceLastRateLimit = 0;
```

Comment:
As this variable could be adjusted by different threads, we must ensure thread-safety here.
```suggestion
    private AtomicLong bytesSinceLastRateLimit = new AtomicLong();
```

Reply:
I think it actually can be even simpler: bytes counter is relevant per-analyse, idk why I declared it as a class field in the first place. I use it now per sampling in https://github.com/crate/crate/pull/13046/commits/ef4e45b308401b4851e502b5a1d80b3b1d6d24d1.

Updated counter name as well.